modules/core/04-channel/keeper/upgrade.go

```diff
@@ -68,3 +72,158 @@ func (k Keeper) constructProposedUpgrade(ctx sdk.Context, portID, channelID stri
 		LatestSequenceSend: seq - 1,
 	}, nil
 }
+
+// ChanUpgradeTry is called by a module to accept the first step of a channel upgrade
+// handshake initiated by a module on another chain. If this function is successful, the upgrade sequence
+// will be returned. If an error occurs in the callback, 0 will be returned but the upgrade sequence will
+// be incremented.
+func (k Keeper) ChanUpgradeTry(
+	ctx sdk.Context,
+	portID,
+	channelID string,
+	proposedUpgradeFields types.UpgradeFields,
+	proposedUpgradeTimeout types.UpgradeTimeout,
+	counterpartyProposedUpgrade types.Upgrade,
+	counterpartyUpgradeSequence uint64,
+	proofCounterpartyChannel,
+	proofUpgrade []byte,
+	proofHeight clienttypes.Height,
+) (proposedUpgrade types.Upgrade, err error) {
+	channel, found := k.GetChannel(ctx, portID, channelID)
+	if !found {
+		return types.Upgrade{}, errorsmod.Wrapf(types.ErrChannelNotFound, "port ID (%s) channel ID (%s)", portID, channelID)
+	}
+
+	// the channel state could be in INITUPGRADE if we are in a crossing hellos situation
+	if !collections.Contains(channel.State, []types.State{types.OPEN, types.INITUPGRADE}) {
+		return types.Upgrade{}, errorsmod.Wrapf(types.ErrInvalidChannelState, "expected one of [%s, %s], got %s", types.OPEN, types.INITUPGRADE, channel.State)
+	}
+
+	// validate the proposed upgrade fields against the existing channel
+	if err = k.ValidateUpgradeFields(ctx, proposedUpgradeFields, channel); err != nil {
+		return types.Upgrade{}, errorsmod.Wrapf(types.ErrInvalidUpgrade, "proposed upgrade fields are invalid: %s", err.Error())
+	}
+
+	connectionEnd, err := k.GetConnection(ctx, channel.ConnectionHops[0])
+	if err != nil {
+		return types.Upgrade{}, err
+	}
+
+	counterpartyConnectionHops := []string{connectionEnd.GetCounterparty().GetConnectionID()}
+
+	expectedCounterpartyChannel := types.Channel{
+		State:           types.INITUPGRADE,
+		Counterparty:    types.NewCounterparty(channel.Counterparty.PortId, channel.Counterparty.ChannelId),
+		Ordering:        channel.Ordering,
+		ConnectionHops:  counterpartyConnectionHops,
+		Version:         channel.Version,
```

Comment:
Do we need to provide the counter party version here too? In the case asymmetric versions this value will not be correct.

Reply:
this looks to not be the case, but this value should be coming from the application callback from the INIT step.

Reply:
i guess this value should actually be from the counterparty proposed upgrade, since it's the expected counterparty channel?

Reply:
I think the method signature needs to be added to include this

Reply:
i think it's already in the proposed upgrade as per the new refactor

Reply:
No it shouldn't be. Per the new refactor, the channel stays with the original parameters until we finally go to OPEN and then we change the parameters.

So in this step, the version in the channel path should be the last successfully negotiated version